src/components/graph/ExerciseTreeGraph.tsx

```diff
@@ -0,0 +1,496 @@
+      .attr('x', '50%')
+      .attr('y', 45)
+      .attr('text-anchor', 'middle')
+
+    /**
+     * Draw show legend hover
+     */
+
+    const showLegend = graph.append('g').attr('class', 'show-legend')
+
+    showLegend
+      .append('rect')
+      .attr('x', 15)
+      .attr('y', 15)
+      .attr('height', 20)
+      .attr('width', 20)
+      .style('fill', '#85C1E9')
+      .style('stroke', 'black')
+      .style('stroke-width', 1)
+
+    showLegend
+      .append('text')
+      .style('font-size', 16)
+      .style('fill', '#EAF2F8')
+      .text('ℹ')
+      .attr('x', 25)
+      .attr('y', 31)
+      .attr('text-anchor', 'middle')
+      .attr('pointer-events', 'none')
+
+    /**
+     * Draw legend
+     */
+    const legend = graph
+      .append('g')
+      .attr('class', 'legend')
+      .attr('visibility', 'hidden')
+
+    legend
+      .append('rect')
+      .attr('x', 1)
+      .attr('y', 1)
+      .attr('height', 70)
+      .attr('width', 225)
+      .style('fill', 'white')
+      .style('stroke', 'black')
+      .style('stroke-width', 1)
+
+    legend
+      .append('circle')
+      .attr('r', circleRadius)
+      .attr('cx', 20)
+      .attr('cy', 20)
+      .style('fill', 'lightsteelblue')
+      .style('stroke', 'black')
+      .style('stroke-width', 1)
+
+    legend
+      .append('rect')
+      .attr('width', squareLength)
+      .attr('height', squareLength)
+      .attr('rx', squareCornerRadius)
+      .attr('ry', squareCornerRadius)
+      .attr('x', 20 - squareLength / 2)
+      .attr('y', 50 - squareLength / 2)
+      .style('fill', 'lightpink')
+      .style('stroke', 'black')
+      .style('stroke-width', 1)
+
+    legend
+      .append('text')
+      .style('font-size', 14)
+      .style('fill', '#000')
+      .text('implemented exercises')
+      .attr('x', 35)
+      .attr('y', 25)
+
+    legend
+      .append('text')
+      .style('font-size', 14)
+      .style('fill', '#000')
+      .text('concepts missing an exercise')
+      .attr('x', 35)
+      .attr('y', 55)
+
+    legend
+      .append('rect')
+      .attr('x', 1)
+      .attr('y', 1)
+      .attr('height', 70)
+      .attr('width', 225)
+      .attr('class', 'overlay')
+      .style('fill', 'rgba(0,0,0,0)')
+
+    showLegend.on('mouseover', () => {
+      d3.select('g.legend').attr('visibility', 'visible')
+    })
+
+    d3.select('g.legend > rect.overlay').on('mouseout', () => {
+      d3.select('g.legend').attr('visibility', 'hidden')
+    })
+  }
+
+  public componentWillUnmount(): void {
+    const circles = Array.from(
+      document.querySelectorAll<SVGCircleElement>('svg#concept-map circle')
+    )
+    for (const circle of circles) {
+      circle.removeEventListener('mouseover', handleCircleMouseover)
+      circle.removeEventListener('mouseout', handleCircleMouseout)
+    }
+
+    d3.select('g.show-legend').on('mouseover', null)
+    d3.select('g.legend > rect.overlay').on('mouseout', null)
+  }
```

Comment:
This is great. You can probably get away with not doing this, at all, but you need to check. So if these event listeners are actually bound to something inside `#graph-container`, they will be cleaned up automagically when `graph-container` unmounts.

Check out [this great post](https://jakearchibald.com/2020/events-and-gc/).

Reply:
I used the d3 `on` function to set the mouse event is that it handles all of the bubbling over SVG elements (which do not bubble like other DOM elements).  So I am not sure if it is necessary or not.

Reply:
Really interesting post though! Really makes the abort controller straight forward